pkg/virt-handler/rest/console.go

```diff
@@ -0,0 +1,195 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2017 Red Hat, Inc.
+ *
+ */
+
+package rest
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"path"
+	"strconv"
+	"sync"
+
+	"github.com/emicklei/go-restful"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
+
+	v1 "kubevirt.io/client-go/api/v1"
+	"kubevirt.io/client-go/kubecli"
+	"kubevirt.io/client-go/log"
+	"kubevirt.io/kubevirt/pkg/virt-handler/isolation"
+)
+
+type ConsoleHandler struct {
+	podIsolationDetector isolation.PodIsolationDetector
+	serialStopChans      map[types.UID](chan struct{})
+	vncStopChans         map[types.UID](chan struct{})
+	serialLock           *sync.Mutex
+	vncLock              *sync.Mutex
+	vmiInformer          cache.SharedIndexInformer
+}
+
+func NewConsoleHandler(podIsolationDetector isolation.PodIsolationDetector, vmiInformer cache.SharedIndexInformer) *ConsoleHandler {
+	return &ConsoleHandler{
+		podIsolationDetector: podIsolationDetector,
+		serialStopChans:      make(map[types.UID](chan struct{})),
+		vncStopChans:         make(map[types.UID](chan struct{})),
+		serialLock:           &sync.Mutex{},
+		vncLock:              &sync.Mutex{},
+		vmiInformer:          vmiInformer,
+	}
+}
+
+func (t *ConsoleHandler) VNCHandler(request *restful.Request, response *restful.Response) {
+	vmi, err := t.getVMI(request)
+	if err != nil {
+		log.Log.Reason(err).Error("Failed retrieve VMI")
+		response.WriteError(http.StatusBadRequest, err)
+		return
+	}
+	unixSocketPath, err := t.getUnixSocketPath(vmi, "virt-vnc")
+	if err != nil {
+		log.Log.Reason(err).Error("Failed unix socket for VNC console")
+		response.WriteError(http.StatusBadRequest, err)
+		return
+	}
+	uid := vmi.GetUID()
+	stopChn := newStopChan(uid, t.vncLock, t.vncStopChans)
+	cleanup := func() {
+		deleteStopChan(uid, stopChn, t.vncLock, t.vncStopChans)
+	}
+	t.stream(request, response, unixSocketPath, stopChn, cleanup)
+}
+
+func (t *ConsoleHandler) SerialHandler(request *restful.Request, response *restful.Response) {
+	vmi, err := t.getVMI(request)
+	if err != nil {
+		log.Log.Reason(err).Error("Failed retrieve VMI")
+		response.WriteError(http.StatusBadRequest, err)
+		return
+	}
+	unixSocketPath, err := t.getUnixSocketPath(vmi, "virt-serial0")
```

Comment:
You will have to check if the path exists. If it does not exist, consult the domain cache and see if the VMI was reported as down, or as not existent, to return appropriate errors. In theory we should probably even consolidate with virt-handler and not rely on the domain cache, but that would be too much for the PR. We can deal with this very small time window later. Looking into the domain cache is good  enough for now.